### PR TITLE
Generate correct debug info for deoptimization target methods

### PR DESCRIPTION
**Pull request by adinn**
This is a proposed fix for issue #2587. It handles the case where deopt target methods are included in a native image. Compiled code for these methods is generated after the compiled code for normal compiled methods . In consequence, a class which contains deopt targets has code occupying two distinct code cache address ranges and code form different classes may be interleaved.

This is a problem because DWARF debug data (info, ranges, frame) must be generated as a sequence of entries associated with non-overlapping address ranges ordered low to high. So, a class containing deopt target methods must be processed twice, generating DWARF entries first for the normal methods and then for the deopt targets.

### CODE REVIEW

File: substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/Range.java [hunk 1]
```diff
@@ -133,6 +133,10 @@ public String getFullMethodName() {
         return fullMethodName;
     }
 
+    public boolean isDeoptTarget() {
+        return methodName.endsWith("**");
```
Comment (olpaw) [inline substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/Range.java:137]: Might be a good idea to define a public static field `METHOD_NAME_DEOPT_SUFFIX` for `"**"` in `src/com.oracle.svm.hosted/src/com/oracle/svm/hosted/meta/HostedMethod.java:247` and use that field also here.

Comment (adinn) [reply substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/Range.java:137]: Yes, will do.

**Comment by adinn** [comment]
Hi @olpaw Just pushed a fix for the rework. I also corrected a problem setting the compile unit index for the deopt target methods compile unit (this was not caught by previous testing because assertions were not enabled when I built libployglot).

This patch now allows @YaSuenag's  libpolyglot crasherto be debugged. Also an objdump of all the debug sections produces no warnings or errors and an eyeball scan shows everything generated with the right values and in the right order.

**Comment by olpaw** [comment]
@adinn thanks a lot. I will push it through out gates.

**Comment by YaSuenag** [comment]
Thanks @adinn ! I got the following backtrace on GDB with your patch:

```
(gdb) bt
#0  com.oracle.svm.core.jdk.VMErrorSubstitutions::doShutdown(java.lang.String, java.lang.Throwable)(void) () at com/oracle/svm/core/log/RealLog.java:385
#1  0x00007f0a2e9ad0c9 in com.oracle.svm.core.util.VMError::shouldNotReachHere(java.lang.String, java.lang.Throwable)(void) () at com/oracle/svm/core/threadlocal/FastThreadLocalInt.java:51
#2  0x00007f0a2e9ad05c in com.oracle.svm.core.util.VMError::guarantee(boolean, java.lang.String)(void) () at com/oracle/svm/core/util/VMError.java:88
#3  0x0000000000000000 in ?? ()
```

I wonder why frame # 3 could not be shown. It seems to lack of DWARF for C API in libpolyglot.so.

**Comment by YaSuenag** [comment]
In addition, RealLog.java:385 does not contain VMErrorSubstitutions::doShutdown.

```
383     @Override
384     public Log autoflush(boolean onOrOff) {
385         autoflush = onOrOff;
386         return this;
387     }
```

**Comment by adinn** [comment]
Hi @YaSuenag 

    (gdb) bt
    #0  com.oracle.svm.core.jdk.VMErrorSubstitutions::doShutdown(java.lang.String, java.lang.Throwable)(void) () at com/oracle/svm/core/log/RealLog.java:385
    #1  0x00007f0a2e9ad0c9 in com.oracle.svm.core.util.VMError::shouldNotReachHere(java.lang.String, java.lang.Throwable)(void) () at com/oracle/svm/core/threadlocal/FastThreadLocalInt.java:51
    #2  0x00007f0a2e9ad05c in com.oracle.svm.core.util.VMError::guarantee(boolean, java.lang.String)(void) () at com/oracle/svm/core/util/VMError.java:88
    #3  0x0000000000000000 in ?? ()

Yes, that is the same stack backtrace I saw.

> I wonder why frame # 3 could not be shown. It seems to lack of DWARF for C API in libpolyglot.so.

This looks like it may be a separate issue to do with the frame section data for frame #2. The stack unwind has not found a valid return address. I'll investigate that as a follow-up to this fix.

> In addition, RealLog.java:385 does not contain VMErrorSubstitutions::doShutdown.

    383     @Override
    384     public Log autoflush(boolean onOrOff) {
    385         autoflush = onOrOff;
    386         return this;
    387     }

I think you are reading that the wrong way round. The frame info indicates that this is a frame for compiled method VMErrorSubstitutions::doShutdown and that the current code address was derived from source code in file RealLog.java. So, that means the current code address is inlined code and you need to look for the call to autoflush in (or below) method VMErrorSubstitutions::doShutdown. And indeed at line 100 of VMErrorSubstitutions.java we have

    100    private static void doShutdown(String msg, Throwable ex) {
    101        try {
    102            Log log = Log.log();
    103            log.autoflush(true);

So, this appears to be correct.

**Comment by olpaw** [comment]
> @adinn thanks a lot. I will feed it to our CI.

Gates all pass. Except for that style issue:
```
1. ERROR in /b/b/e/main/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfInfoSectionImpl.java (at line 215)
	private int findLo(LinkedList<PrimaryEntry> classPrimaryEntries, boolean isDeoptTargetCU) {
	            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The method findLo(LinkedList<PrimaryEntry>, boolean) from the type DwarfInfoSectionImpl can be declared as static
----------
2. ERROR in /b/b/e/main/substratevm/src/com.oracle.objectfile/src/com/oracle/objectfile/elf/dwarf/DwarfInfoSectionImpl.java (at line 233)
	private int findHi(LinkedList<PrimaryEntry> classPrimaryEntries, boolean includesDeoptTarget, boolean isDeoptTargetCU) {
	            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
The method findHi(LinkedList<PrimaryEntry>, boolean, boolean) from the type DwarfInfoSectionImpl can be declared as static
----------
2 problems (2 errors)
```
Please fix.

**Comment by adinn** [comment]
<sigh/> Should be ok now :-)

**Comment by olpaw** [comment]
With https://github.com/oracle/graal/pull/2612/commits/6c5420a33a06c782d709e48165bf6b89cc442981 `mx eclipseformat` fails
```
Changes:
--- a/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/Range.java
+++ b/src/com.oracle.objectfile/src/com/oracle/objectfile/debugentry/Range.java
@@ -55,7 +55,8 @@
     /*
      * Create a primary range.
      */
-    public Range(String fileName, Path filePath, String className, String methodName, String paramNames, String returnTypeName, StringTable stringTable, int lo, int hi, int line, boolean isDeoptTarget) {
+    public Range(String fileName, Path filePath, String className, String methodName, String paramNames, String returnTypeName, StringTable stringTable, int lo, int hi, int line,
+                    boolean isDeoptTarget) {
         this(fileName, filePath, className, methodName, paramNames, returnTypeName, stringTable, lo, hi, line, isDeoptTarget, null);
     }
 
@@ -65,10 +66,12 @@
     public Range(String fileName, Path filePath, String className, String methodName, String paramNames, String returnTypeName, StringTable stringTable, int lo, int hi, int line, Range primary) {
         this(fileName, filePath, className, methodName, paramNames, returnTypeName, stringTable, lo, hi, line, false, primary);
     }
+
     /*
      * Create a primary or secondary range.
      */
-    private Range(String fileName, Path filePath, String className, String methodName, String paramNames, String returnTypeName, StringTable stringTable, int lo, int hi, int line, boolean isDeoptTarget, Range primary) {
+    private Range(String fileName, Path filePath, String className, String methodName, String paramNames, String returnTypeName, StringTable stringTable, int lo, int hi, int line,
+                    boolean isDeoptTarget, Range primary) {
         /*
          * Currently file name and full method name need to go into the debug_str section other
          * strings just need to be deduplicated to save space.
```

**Comment by adinn** [comment]
Hmm, that didn't happen with my local eclipseformat. I am using 4.14version

**Comment by adinn** [comment]
Hi @YaSuenag

I have identified why stack frame #3 has not been correctly unwound. It's a problem in the frame data being passed to through to the debug info generator which will need a compiler change to fix it. I'll raise a separate issue for it.

**Comment by olpaw** [comment]
> Hmm, that didn't happen with my local eclipseformat. I am using 4.14. Is the build using a later one? I'll correct this by hand.

Our `gate-vm-style-linux-amd64` shows
```
> 2020-06-26 14:41:11(+00:09) [INFO] Set environment variable ECLIPSE_EXE=/sc/bb/downloads/eclipse-4.14.0-linux-amd64_8aabce73d3/eclipse/eclipse
```
in its logs.

**Pull request merged**